# DS-SF-38 | 02 | Python | Assignment | Starter Code

## Python

> ## "Data scientists' most basic, universal skill is the ability to write code."
>
> Harvard Business Review | “Data Scientists: The Sexiest Job of the 21st Century” (2012)
>
> (https://hbr.org/2012/10/data-scientist-the-sexiest-job-of-the-21st-century)

Today, to give you some more practice beyond the course pre-requisites, we are going to implement a few functions in Python.  This assignment covers to some extend the following topics:

- Functions (defining and using your own functions but also calling functions from packages)
- Loops
- Arithmetic operations
- Conditional statements

**Don't worry if you get stuck.  Ask around, review the answer key, and ask around more.  As this course progresses, your programming proficiency will increase.**

> ### Question 1.  Multiples of 3 and 5
>
> If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.  The sum of these multiples is 23.
>
> Find the sum of all the multiples of 3 or 5 below 1,000.
>
> (Source: [Project Euler | Problem 1](https://projecteuler.net/problem=1))

In [17]:
# TODO 
# 9/1 6:45-6:55pm
def get_sum_multiples_3_5(val):
    #initialize sum_multiples
    sum_multiples = 0
    #loop through 1 to val
    for x in range(val):
        # get multiples of 3, 5, if % 0 add to sum
        if x % 3 == 0 or x % 5 == 0:
            sum_multiples = sum_multiples + x

    print(sum_multiples)
    return sum_multiples
   
get_sum_multiples_3_5(1000)


# Ivan's answer
def multiples(n):
    accumulator = 0
    for i in range(1, n):
        if (i % 3 == 0) or (i %  5 == 0):
            accumulator += i
    return accumulator

# multiples(1000)

233168

Answer: 233168

> ### Question 2.  Estimating square roots
>
> Given a real number $m$, let's define the series $u$ as follow:
> - $u_0 = 1$
> - $u_{n+1} = \frac{u_n ^ 2 + m}{2u_n}$
>
>
> Implement the calculations of the series $u$ above to estimate square roots.  Verify that $\sqrt{144} = 12$ and use your function to calculate $\sqrt{1024}$.

In [30]:
# TODO
# 9/1 7:04 - 7/14
# def sqrt(m):
#     u_n = 1.
    
#     while True:
#         u_n_plus_1 = (u_n ** 2 + m) / (2 * u_n)
        
#         # @TODO don't get why the comparator
#         if abs(u_n_plus_1 / u_n - 1)
        
#     return u_n
    
#########################
# Ivan's code 

def sqrt(m):
    u_n = 1.

    while True:
        u_n_plus_1 = (u_n ** 2 + m) / (2 * u_n)

        if abs(u_n_plus_1 / u_n - 1) < 10 ** -6:
            break

        u_n = u_n_plus_1

    return u_n
    

Answer: TODO

> ### Question 3.  Prime Numbers
>
> A prime (number) is a natural number greater than 1 that has no positive divisors other than 1 and itself.  ([Wikipedia](https://en.wikipedia.org/wiki/Prime_number))
>
> Calculate all primes below 1,000.  What's their sum?

In [40]:
# TODO
# 9/1 7:15 - 7:35

# For this prime function, ended up looking @ ivans
def is_prime(n):
    for i in range (2, int(sqrt(n)+1)): # oh okay, the sqrt is so it doesnt have to loop all the way up to n, but it wouldn't even hit it if returned False at any of the integers by the halfway point
        if n % i == 0:
            return False
    # this catches if n == 1, 2
    return True

def find_primes(val):
    primes_list = []
    for i in range(2, val+1):
        # if i is a prime put it in the list
        if is_prime(i):
#             print i
            primes_list.append(i)
    print primes_list
    return primes_list

primes_list = find_primes(100)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


In [41]:
sum(primes_list)

1060

Answer: TODO

> ### Question 4.  Largest prime factor
>
> The prime factors of 13195 are 5, 7, 13 and 29.
>
> What is the largest prime factor of the number 600851475143?
>
> (Source: [Project Euler | Problem 3](https://projecteuler.net/problem=3))

In [ ]:
# TODO
#

Answer: TODO

> ### Question 5.  Mean
>
> Write a function to calculate the mean (average) of a list.
>
> What's the mean of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [ ]:
# TODO

Answer: TODO

> ### Question 6.  Sample standard deviation
>
> Write a function to calculate the standard deviation of a sample.
>
> Given the sample $x_1, x_2, ..., x_N$, its standard deviation is defined as $s = \sqrt{\frac{1}{N - 1} \sum_{i = 1}^{N} (x_i  - \bar{x})^2}$, with $\bar{x}$ as the sample mean.
>
> What's the standard deviation of the following sample: 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?
>
> ([Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation))

In [ ]:
# TODO

Answer: TODO

> ### Question 7.  Median
>
> Write a function to calculate the median ("middle value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Median))
>
> What's the median of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [6]:
# TODO
# 9/4 9:30am - 9:40
def get_median(num_arr):
    
    med_val = 0
    # Order list first
    num_arr.sort()
    print(num_arr)
    
    # val @ mid length
    med_val = num_arr[len(num_arr)/2]
    
    return med_val


# Ivan's version
def median(xs):
    n = len(xs) # lots of logic around the length
    if n == 0:  # Okay I forgot edge cases
        return None
    sorted_xs = sorted(xs)
    
    i = n / 2 # middle
    if(n % 2 == 1):
        return sorted_xs[i]
    else:
        return mean(sorted_xs[i - 1: i + 1]) # For even #d lists

test_arr = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]
get_median(test_arr)
median(test_arr)

[4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]


9

Answer: TODO

> ### Question 8.  Mode
>
> Write a function to calculate the mode ("most frequent value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Mode_(statistics)))
>
> What's the mode of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7 and 5?  How about the mode of 8, 8, 8, 8, 8, 8, 19, 8, 8 and 8?

In [12]:
# TODO
# 9/4 9:50 - 10:00
# Too complex, dictionary approach

def get_mode(num_arr):
    # sort first
    sorted_arr = sorted(num_arr)
    # loop through, if val is same as the one just now, that num is now the 
    # build dict, seems excessive, def would refactor
    d = {}
    for i in sorted_arr:
        if sorted_arr[i] == sorted_arr[i-1]:
#             d[]
    
    
    # if the num count is greater than half the ize of arr, DEF most frequent
    return []

#########
# Ivan's
#########


test_arr = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]
# sorted(test_arr)
get_mode(test_arr)

    

4
5
6
7
8
9
10
11
12
13
14


[]

Answer: TODO